### PR TITLE
#297: Decouple UMA-PEP into basic level

### PR DESCRIPTION
**Pull request by ldeveloperl1985**
Just decouple UMA-PEP into basic level to get existing functionality
@altexy don't merge it, just review it

### CODE REVIEW

File: kong/common/kong-auth-pep-common.lua [hunk 1]
```diff
@@ -234,6 +232,9 @@ local function access_granted(conf, token_data)
         [const.CONSUMER_USERNAME] = tostring(consumer.username),
     }
 
+    kong.ctx.shared.authenticated_consumer = token_data.consumer -- forward compatibility
+    ngx.ctx.authenticated_consumer = token_data.consumer -- backward compatibility
```
Comment (altexy) [inline kong/common/kong-auth-pep-common.lua]: Kong's plugin don't store `authenticated_consumer` for anonymous use case.
Later we may distinguish anonymous use case by missing `authenticated_consumer`

File: kong/common/kong-auth-pep-common.lua [hunk 2]
```diff
@@ -328,18 +323,14 @@ function hooks.is_access_granted(self, conf, protected_path, method, scope_expre
 end
  ]]
 
-_M.access_handler = function(self, conf, hooks)
+_M.access_uma_handler = function(self, conf, hooks)
     local authorization = ngx.var.http_authorization
-    local token = get_token(authorization)
+    local token = get_token(authorization) or kong.ctx.shared.request_token
```
Comment (altexy) [inline kong/common/kong-auth-pep-common.lua]: Don't need token here, take a look at my comment above.

File: kong/common/kong-auth-pep-common.lua [hunk 3]
```diff
@@ -359,6 +350,68 @@ _M.access_handler = function(self, conf, hooks)
...
+    local token_data = kong.ctx.shared.authenticated_token
+    if not token_data then
+        return kong.response.exit(401, { message = "Token not authenticated" })
+    else
+        client_id = token_data.client_id
+        exp = token_data.exp
+    end
+
+    if not protected_path then
+        return access_granted(conf, token_data)
```
Comment (altexy) [inline kong/common/kong-auth-pep-common.lua]: Don't need to call `access_granted` It is only to be used now from auth. plugins.

File: kong/common/kong-auth-pep-common.lua [hunk 4]
```diff
@@ -359,6 +350,68 @@ _M.access_handler = function(self, conf, hooks)
...
+    end
+
+    local cache_key, pending = hooks.build_cache_key(method, protected_path, token, token_data.scope)
+
+    if cache_key then
+        local is_access_granted = worker_cache_get_pending(cache_key)
+        if is_access_granted then
+            kong.ctx.shared[self.metric_client_granted] = true
+            return access_granted(conf, token_data)
```
Comment (altexy) [inline kong/common/kong-auth-pep-common.lua]: same

File: kong/common/kong-auth-pep-common.lua [hunk 5]
```diff
@@ -359,6 +350,68 @@ _M.access_handler = function(self, conf, hooks)
...
+
+    if pending then
+        set_pending_state(cache_key)
+    end
+    if hooks.is_access_granted(self, conf, protected_path, method, scope_expression, token_data.scope, token) then
+        if cache_key then
+            worker_cache:set(cache_key, true, exp - ngx.now() - EXPIRE_DELTA)
+        end
+        kong.ctx.shared[self.metric_client_granted] = true
+        return access_granted(conf, token_data)
```
Comment (altexy) [inline kong/common/kong-auth-pep-common.lua]: same

File: kong/common/kong-auth-pep-common.lua [hunk 6]
```diff
@@ -359,6 +350,68 @@ _M.access_handler = function(self, conf, hooks)
 
     kong.log.debug("protected resource path: ", protected_path, " URI: ", path, " token: ", token)
 
+    local client_id, exp
+    local token_data = kong.ctx.shared.authenticated_token
+    if not token_data then
+        return kong.response.exit(401, { message = "Token not authenticated" })
```
Comment (altexy) [inline kong/common/kong-auth-pep-common.lua]: IMO authorization plugin must always respond with 403

File: kong/common/kong-auth-pep-common.lua [hunk 7]
```diff
@@ -359,6 +350,68 @@ _M.access_handler = function(self, conf, hooks)
...
+Authentication
+
+@return introspect_response, status, err
+upon success returns only introspect_response,
+otherwise return nil, status, err
+function hooks.introspect_token(self, conf, token)
+end
+
+ ]]
+_M.access_auth_handler = function(self, conf, hooks)
```
Comment (altexy) [inline kong/common/kong-auth-pep-common.lua]: hooks for this common code always contains one function - we may pass it directly

File: kong/common/kong-auth-pep-common.lua [hunk 8]
```diff
@@ -408,46 +459,19 @@ _M.access_handler = function(self, conf, hooks)
...
-        )
+            exp - ngx.now() - EXPIRE_DELTA)
     else
         client_id = token_data.client_id
         exp = token_data.exp
     end
 
     -- Client(Consumer) is authenticated
-    kong.ctx.shared.authenticated_consumer = token_data.consumer -- forward compatibility
-    ngx.ctx.authenticated_consumer = token_data.consumer -- backward compatibility
```
Comment (altexy) [inline kong/common/kong-auth-pep-common.lua]: should be restored

File: kong/plugins/gluu-uma-auth/access.lua [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+    local introspect_rpt_response_data = try_introspect_rpt(conf, token, ptoken)
+    if not introspect_rpt_response_data.active then
+        return nil, 401, "Invalid access token provided in Authorization header"
+    end
+    return introspect_rpt_response_data
+end
+
+return function(self, conf)
+    kong_auth_pep_common.access_auth_handler(self, conf, hooks)
```
Comment (altexy) [inline kong/plugins/gluu-uma-auth/access.lua]: pass `introspect_token` instead of table

File: kong/plugins/gluu-uma-pep/schema.lua [hunk 10]
```diff
@@ -17,9 +17,7 @@ return {
         op_url = { required = true, type = "url" },
         uma_scope_expression = { required = true, func = check_expression, type = "table" },
         deny_by_default = { type = "boolean", default = true },
-        anonymous = { type = "string", func = kong_auth_pep_common.check_user, default = "" },
-        hide_credentials = { type = "boolean", default = false },
-        ignore_scope = { type = "boolean", default = false }
+        anonymous = { type = "string", func = kong_auth_pep_common.check_user, default = "" }
```
Comment (altexy) [inline kong/plugins/gluu-uma-pep/schema.lua]: do we really need it?

**Comment by altexy** [comment]
I didn't review sequence diagrams so far.

File: kong/common/kong-auth-pep-common.lua [hunk 11]
```diff
@@ -328,18 +323,14 @@ function hooks.is_access_granted(self, conf, protected_path, method, scope_expre
 end
  ]]
 
-_M.access_handler = function(self, conf, hooks)
+_M.access_uma_handler = function(self, conf, hooks)
```
[the remark this answers is not in the document]
Comment (ldeveloperl1985) [reply kong/common/kong-auth-pep-common.lua]: ah yes right. I'll change it

File: kong/common/kong-auth-pep-common.lua (hunk 3 above)
Comment (ldeveloperl1985) [reply kong/common/kong-auth-pep-common.lua]: Agreed

File: kong/common/kong-auth-pep-common.lua (hunk 6 above)
Comment (ldeveloperl1985) [reply kong/common/kong-auth-pep-common.lua]: Agreed

File: kong/common/kong-auth-pep-common.lua (hunk 7 above)
Comment (ldeveloperl1985) [reply kong/common/kong-auth-pep-common.lua]: Yes, I'll follow this

File: kong/plugins/gluu-uma-pep/schema.lua (hunk 10 above)
Comment (ldeveloperl1985) [reply kong/plugins/gluu-uma-pep/schema.lua]: I think no. I'll remove this

File: kong/common/kong-auth-pep-common.lua (hunk 1 above)
Comment (ldeveloperl1985) [reply kong/common/kong-auth-pep-common.lua]: Yes but then how to handle the ticket case when admin configure uma-pep for specific consumer.

File: kong/common/kong-auth-pep-common.lua (hunk 2 above)
Comment (ldeveloperl1985) [reply kong/common/kong-auth-pep-common.lua]: We need it for to check access and set cache

File: kong/common/kong-auth-pep-common.lua [hunk 12]
```diff
@@ -419,35 +461,11 @@ _M.access_handler = function(self, conf, hooks)
...
-        if cache_key then
-            worker_cache:set(cache_key, true, exp - ngx.now() - EXPIRE_DELTA)
-        end
-        kong.ctx.shared[self.metric_client_granted] = true
-        return access_granted(conf, token_data)
-    end
-    if pending then
-        clear_pending_state(token)
+    kong.ctx.shared.authenticated_token = token_data -- Used to check wether token is authenticated or not for PEP plugin
+    if conf.hide_credentials then
```
Comment (altexy) [inline kong/common/kong-auth-pep-common.lua]: IMO we must pass a token independent from `hide_credentials` value

File: kong/common/kong-auth-pep-common.lua (hunk 2 above)
Comment (altexy) [reply kong/common/kong-auth-pep-common.lua]: it should be just ` local token = kong.ctx.shared.request_token`

File: kong/common/kong-auth-pep-common.lua [hunk 13]
```diff
@@ -381,10 +426,8 @@ _M.access_handler = function(self, conf, hooks)
...
 
-            if not protected_path then
-                if #conf.anonymous > 0 then
-                    return handle_anonymous(conf)
-                end
+            if #conf.anonymous > 0 then
+                return handle_anonymous(conf)
             end
 
             return kong.response.exit(401, { message = "Bad token" })
```
Comment (altexy) [inline kong/common/kong-auth-pep-common.lua]: `message = err`

File: kong/common/kong-auth-pep-common.lua [hunk 14]
```diff
@@ -327,19 +319,14 @@ function build_cache_key(method, protected_path, token, scopes)
...
 
     local method = ngx.req.get_method()
     local path = ngx.var.uri
     local protected_path, scope_expression
-    if not conf.ignore_scope then
-        protected_path, scope_expression = hooks.get_path_by_request_path_method(
-            self, conf, path, method
-        )
-    end
+    protected_path, scope_expression = hooks.get_path_by_request_path_method(self, conf, path, method)
```
[the remark this answers is not in the document]
Comment (ldeveloperl1985) [reply kong/common/kong-auth-pep-common.lua]: Okay

File: kong/common/kong-auth-pep-common.lua (hunk 12 above)
Comment (ldeveloperl1985) [reply kong/common/kong-auth-pep-common.lua]: Agreed

File: kong/common/kong-auth-pep-common.lua (hunk 13 above)
Comment (ldeveloperl1985) [reply kong/common/kong-auth-pep-common.lua]: Okay, I'll update it

File: kong/common/kong-auth-pep-common.lua (hunk 2 above)
Comment (ldeveloperl1985) [reply kong/common/kong-auth-pep-common.lua]: Okay

File: kong/common/kong-auth-pep-common.lua [hunk 15]
```diff
@@ -321,12 +321,11 @@ end
  ]]
 _M.access_pep_handler = function(self, conf, hooks)
     local authorization = ngx.var.http_authorization
```
Comment (altexy) [inline kong/common/kong-auth-pep-common.lua]: We don't need it anymore

Comment (ldeveloperl1985) [reply kong/common/kong-auth-pep-common.lua]: Okay, I'll remove this

**Comment by ldeveloperl1985** [comment]
> OK, what is the state of our test suite?

I've written test cases for all the cases. In AUTH case, there is test case related to only Auth plugin. In PEP case, I test with both plugin : AUTH + PEP

**Comment by altexy** [comment]
What is the target branch for this PR?

**Comment by ldeveloperl1985** [comment]
> What is the target branch for this PR?

Master branch. I think, we can work in master and after next release, will make a new branch from master with related version.

**Pull request merged**